Add regions2d masking 2d to 3d to 2d functionality and Regions2d Masking Simplification

ctuguinay opened the pull request:
### Regions Mask conversion between 2D and 3D and vice-versa

This PR addresses #92.

- Added functionality for turning 2d region mask to 3d region mask
- Added functionality for turning 3d region mask to 2d region mask
- Created a test with additional "dummy" mask values (without additional values we would have 1 non-nan value) that converted 2d mask to 3d mask to 2d mask and checked for appropriate values at all stages of the test

Attached below is a visual representation of the new features added.
![image](https://github.com/OSOceanAcoustics/echoregions/assets/87830138/629c48bd-4d72-48d2-be65-bdb8223782f2)

###Regions2d Masking Simplification

This PR also addresses #98.

Followed Wu-Jung's instructions for simplifying this code: 

1. Initiated `regionmask.Regions(...)`.
2. Created mask using `r.mask(...)`.
3. In the case where list, converted default labels to mask labels.

valentina-s reviewed (changes requested):
I added a few comments within the code. A few more general comments:

* Add a few tests which particularly deal with cases such as:
- [x] of all nans
- [x] only one label (since then a 3D array may collapse to a 2D array)
- [x] overlapping variables in the 3D mask
- [x] test 3D to 2D to 3D

* I realized  the current version is a bit tuned to going from 2D to 3D, and is limited to the approach we took earlier that 2D masks can have only integer labels, which will result in the 3D masks having only integer labels, unless we expand on the options. In practice we would want to have some labels related to the region, such as "Hake", "Rockfish". So one could consider the option the dictionary to be attached actually to the 2D mask, vs the 3d one, and mapping the integer values to some meaningful class names (which could be coming from the classes in the evr file)?

ctuguinay commented:
Hey @valentina-s,

Thank you for the review!
Based on your comments, I just made a recent push with the following changes:

- Added more concise and more accurate descriptions
- Removed copy from convert mask_2d_to_3d
- Added functionality and notes to work with and denote no masked regions
- Add functionality to check for overlap and throw error if overlap exists
- Created test for 3d overlap
- Created test for 2d-3d-2d-3d (adding additional step to previous existing 2d-3d-2d test)
- Created test for null mask
- Created test for single region mask

As for your following comments:

> Maybe use the dictionary directly instead of the numpy array, since we want to test the mapping, not the order.

The dictionary isn't actually saved as a dictionary in the dataset object. It has to be stored as a numpy array. I'm not exactly sure how to circumvent this, so I just aligned the array so that its values at specific indices aligned with layers that they correspond to at the same indices as the data array in the dataset, i.e. `mask_3d_ds.mask_dictionary.data[0]` value aligns with the layer `mask_3d_ds.mask_3d[0]`.

> I am missing why there is no 11 in the fake values?

The no 11 in there is not in fake values, but it already existed within the original 2d mask. I added 2 fake values here to make sure that we had 3 layers in the 3d mask.

> * I realized the current version is a bit tuned to going from 2D to 3D, and is limited to the approach we took earlier that 2D masks can have only integer labels, which will result in the 3D masks having only integer labels, unless we expand on the options. In practice we would want to have some labels related to the region, such as "Hake", "Rockfish". So one could consider the option the dictionary to be attached actually to the 2D mask, vs the 3d one, and mapping the integer values to some meaningful class names (which could be coming from the classes in the evr file)?

I do agree that having such a "dictionary" (in practice a numpy array) for the 2D mask is important, but I still think having the "dictionary" for 3D should still be there. I'm unfamiliar with the whole labels with Hake and Rockfish and how they may appear in the EVR files since the one that I am currently using doesn't have such labels.

I think this should be addressed in a separate PR since it deals with something that is broader than the 2d and 3d conversion of region masking.

valentina-s commented on echoregions/utils/api.py:
```diff
@@ -2,40 +2,58 @@
+        label in the 2D mask.
+    Notes
+    -----
+    -1 in the dictionary of mask_3d_ds means that there exists no masked
```
I think empty data array would be better: `xr.DataArray(data=None)` and `.isnull()`

valentina-s reviewed (approved):
Thanks for the updates @ctuguinay!  I have added a few small comments to address. In general, there is quite a lot of use of numpy arrays instead of xarray arrays. It would be good to use and test directly the xarrays whenever possible to ensure the labels do not get misaligned.

Regarding the dictionary, if the labels of the 3D masks are automatically named after the integer label for that region in the 2D mask (instead of generic 0,1,2,3....), then one does not need a dictionary. It is ok to set up an issue for this and revisit it once we look at examples with different labels. I will add some details and we can discuss it.

ctuguinay commented:
Hey @valentina-s, thank you for the review!

> Thanks for the updates @ctuguinay! I have added a few small comments to address. In general, there is quite a lot of use of numpy arrays instead of xarray arrays. It would be good to use and test directly the xarrays whenever possible to ensure the labels do not get misaligned.
> 
> Regarding the dictionary, if the labels of the 3D masks are automatically named after the integer label for that region in the 2D mask (instead of generic 0,1,2,3....), then one does not need a dictionary. It is ok to set up an issue for this and revisit it once we look at examples with different labels. I will add some details and we can discuss it.

In my most recent push, I made changes such that the xarray object itself is primarily used. For some of the unique value code, I had to use the inner numpy array since xarray didn't have an equivalent to np.unique.

I also made the addition of a string dictionary a new issue and linked it to this PR.

> I believe `region_ids = r2d.data.region_id.to_list()` achieves the same as those three lines.

In the latest push, I had to modify this into `r2d.data.region_id.astype(float).to_list()`.

> assert also the dimensions: i.e. whether one label gives 3d array or collapses to 2D array

In the latest push, I asserted this.

> this part can be written directly with xarray `where` operations, and avoid looping over the ndindex.

In the latest push, I turned this code into a one-liner using `xr.where`.

> I think empty data array would be better: `xr.DataArray(data=None)` and `.isnull()`

I tried to implement this, but this left the empty case handling a lot less uniform with the code that handled the non-empty case since there did not exist a list anymore to access. Instead, I left the dictionary itself as an empty array and I handled the identification of the empty and non-empty case with len(...).
